# Lab | Pandas

In this lab, we will be working with the customer data from an insurance company, which can be found in the CSV file located at the following link: https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv

The data includes information such as customer ID, state, gender, education, income, and other variables that can be used to perform various analyses.

Throughout the lab, we will be using the pandas library in Python to manipulate and analyze the data. Pandas is a powerful library that provides various data manipulation and analysis tools, including the ability to load and manipulate data from a variety of sources, including CSV files.

### Data Description

- Customer - Customer ID

- ST - State where customers live

- Gender - Gender of the customer

- Education - Background education of customers 

- Customer Lifetime Value - Customer lifetime value(CLV) is the total revenue the client will derive from their entire relationship with a customer. In other words, is the predicted or calculated value of a customer over their entire duration as a policyholder with the insurance company. It is an estimation of the net profit that the insurance company expects to generate from a customer throughout their relationship with the company. Customer Lifetime Value takes into account factors such as the duration of the customer's policy, premium payments, claim history, renewal likelihood, and potential additional services or products the customer may purchase. It helps insurers assess the long-term profitability and value associated with retaining a particular customer.

- Income - Customers income

- Monthly Premium Auto - Amount of money the customer pays on a monthly basis as a premium for their auto insurance coverage. It represents the recurring cost that the insured person must pay to maintain their insurance policy and receive coverage for potential damages, accidents, or other covered events related to their vehicle.

- Number of Open Complaints - Number of complaints the customer opened

- Policy Type - There are three type of policies in car insurance (Corporate Auto, Personal Auto, and Special Auto)

- Vehicle Class - Type of vehicle classes that customers have Two-Door Car, Four-Door Car SUV, Luxury SUV, Sports Car, and Luxury Car

- Total Claim Amount - the sum of all claims made by the customer. It represents the total monetary value of all approved claims for incidents such as accidents, theft, vandalism, or other covered events.


External Resources: https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9

## Challenge 1: Understanding the data

In this challenge, you will use pandas to explore a given dataset. Your task is to gain a deep understanding of the data by analyzing its characteristics, dimensions, and statistical properties.

- Identify the dimensions of the dataset by determining the number of rows and columns it contains.
- Determine the data types of each column and evaluate whether they are appropriate for the nature of the variable. You should also provide suggestions for fixing any incorrect data types.
- Identify the number of unique values for each column and determine which columns appear to be categorical. You should also describe the unique values of each categorical column and the range of values for numerical columns, and give your insights.
- Compute summary statistics such as mean, median, mode, standard deviation, and quartiles to understand the central tendency and distribution of the data for numerical columns. You should also provide your conclusions based on these summary statistics.
- Compute summary statistics for categorical columns and providing your conclusions based on these statistics.

In [10]:
# Your code here

import pandas as pd

url = 'https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv'
data = pd.read_csv (url)

# Dimensions
print("Dataset dimensions:", data.shape)


Dataset dimensions: (4008, 11)


In [11]:
# Data types
  # Convert problematic columns
data['Customer Lifetime Value'] = (
    data['Customer Lifetime Value']
    .astype(str)
    .str.replace('%', '', regex=False)
    .replace(['nan', 'NaN', 'None', ''], pd.NA)
)
data['Customer Lifetime Value'] = pd.to_numeric(data['Customer Lifetime Value'], errors='coerce')

data['Number of Open Complaints'] = (
    data['Number of Open Complaints']
    .astype(str)
    .str.extract(r'(\d+)')
)
data['Number of Open Complaints'] = pd.to_numeric(data['Number of Open Complaints'], errors='coerce')

  # Display data types
print("Data Types:")
print(data.dtypes)
print("-" * 60)

print("Fix Suggestions:")
print("""
- Convert 'Customer Lifetime Value' to numeric 
- Convert 'Number of Open Complaints' to numeric 
- Standardize 'ST' (state), 'GENDER', and 'Education' spelling
""")

Data Types:
Customer                      object
ST                            object
GENDER                        object
Education                     object
Customer Lifetime Value      float64
Income                       float64
Monthly Premium Auto         float64
Number of Open Complaints    float64
Policy Type                   object
Vehicle Class                 object
Total Claim Amount           float64
dtype: object
------------------------------------------------------------
Fix Suggestions:

- Convert 'Customer Lifetime Value' to numeric 
- Convert 'Number of Open Complaints' to numeric 
- Standardize 'ST' (state), 'GENDER', and 'Education' spelling



In [12]:
# Unique values per column
unique_counts = data.nunique()
print("\n Number of Unique Values per Column:")
print(unique_counts)
print("-" * 60)

  # Identify categorical and numerical columns
categorical_cols = data.select_dtypes(include=['object']).columns.tolist()
numerical_cols = data.select_dtypes(include=['number']).columns.tolist()

print("Categorical Columns:", categorical_cols)
print("Numerical Columns:", numerical_cols)
print("-" * 60)


 Number of Unique Values per Column:
Customer                     1071
ST                              8
GENDER                          5
Education                       6
Customer Lifetime Value      1027
Income                        774
Monthly Premium Auto          132
Number of Open Complaints       1
Policy Type                     3
Vehicle Class                   6
Total Claim Amount            761
dtype: int64
------------------------------------------------------------
Categorical Columns: ['Customer', 'ST', 'GENDER', 'Education', 'Policy Type', 'Vehicle Class']
Numerical Columns: ['Customer Lifetime Value', 'Income', 'Monthly Premium Auto', 'Number of Open Complaints', 'Total Claim Amount']
------------------------------------------------------------


In [13]:
#Categorical Column Analysis

print("\nCategorical Column Summary (top unique values):")
for col in categorical_cols:
    print(f"\n{col} (Unique: {data[col].nunique()})")
    print(data[col].value_counts().head(10))

print("-" * 60)


Categorical Column Summary (top unique values):

Customer (Unique: 1071)
Customer
MY31220    1
RB50392    1
QZ44356    1
AI49188    1
WW63253    1
GA49547    1
OC83172    1
XZ87318    1
CF85061    1
DY87989    1
Name: count, dtype: int64

ST (Unique: 8)
ST
Oregon        320
California    211
Arizona       186
Cali          120
Nevada         98
Washington     81
WA             30
AZ             25
Name: count, dtype: int64

GENDER (Unique: 5)
GENDER
F         457
M         413
Male       39
female     28
Femal      17
Name: count, dtype: int64

Education (Unique: 6)
Education
Bachelor                324
College                 313
High School or Below    296
Master                   94
Doctor                   37
Bachelors                 7
Name: count, dtype: int64

Policy Type (Unique: 3)
Policy Type
Personal Auto     780
Corporate Auto    234
Special Auto       57
Name: count, dtype: int64

Vehicle Class (Unique: 6)
Vehicle Class
Four-Door Car    576
Two-Door Car     205
SUV       

In [19]:
#Numerical Column Analysis

# Basic statistics
num_summary = data[numerical_cols].describe().T
num_summary["median"] = data[numerical_cols].median()
num_summary["mode"] = data[numerical_cols].mode().iloc[0]
num_summary["std_dev"] = data[numerical_cols].std()

print("\nNumerical Summary Statistics:")
print(num_summary)
print("-" * 60)

# Range of values
print("\nRange of Each Numerical Column:")
for col in numerical_cols:
    print(f"{col}: {data[col].min()} → {data[col].max()}")

print("-" * 60)

# Insights
print("💡 Insights:")
print("""
- 'Customer Lifetime Value' and 'Monthly Premium Auto' have very large ranges → check for outliers.
- 'Number of Open Complaints' has only one unique value → possible data issue.
- 'Income' and 'Total Claim Amount' look reasonable for customer-level data.
""")


Numerical Summary Statistics:
                            count           mean            std  \
Customer Lifetime Value    1068.0  793690.259345  643478.410278   
Income                     1071.0   39295.701214   30469.427060   
Monthly Premium Auto       1071.0     193.234360    1601.190369   
Number of Open Complaints  1071.0       1.000000       0.000000   
Total Claim Amount         1071.0     404.986909     293.027260   

                                     min            25%            50%  \
Customer Lifetime Value    200435.070000  403407.995000  588174.235000   
Income                          0.000000   14072.000000   36234.000000   
Monthly Premium Auto           61.000000      68.000000      83.000000   
Number of Open Complaints       1.000000       1.000000       1.000000   
Total Claim Amount              0.382107     202.157702     354.729129   

                                   75%           max         median      mode  \
Customer Lifetime Value    896287.1975  

In [20]:
 #Summary for Categorical Columns

print("\nSummary Statistics for Categorical Columns:")
for col in categorical_cols:
    print(f"\n {col}")
    print("Unique Values:", list(data[col].dropna().unique()))
    print("Most Frequent Value(s):")
    print(data[col].value_counts().head(3))
    print("-" * 40)


Summary Statistics for Categorical Columns:

 Customer
Unique Values: ['RB50392', 'QZ44356', 'AI49188', 'WW63253', 'GA49547', 'OC83172', 'XZ87318', 'CF85061', 'DY87989', 'BQ94931', 'SX51350', 'VQ65197', 'DP39365', 'SJ95423', 'IL66569', 'BW63560', 'FV94802', 'OE15005', 'WC83389', 'FL50705', 'ZK25313', 'QK46697', 'YH23384', 'TZ98966', 'HM55802', 'FS42516', 'US89481', 'HS14476', 'GE62437', 'EJ77678', 'SV85652', 'UL64533', 'PF41800', 'HD95276', 'SK67821', 'YV55495', 'KY38074', 'DM79012', 'CM61827', 'WC35801', 'QG25316', 'MB98372', 'IL19217', 'SR38658', 'YD87931', 'HG65722', 'BU27331', 'XM45289', 'KP34198', 'SH90947', 'WE95729', 'PY51963', 'RB69909', 'NW21079', 'FR46645', 'SY17488', 'AP67935', 'FS37417', 'ML29312', 'UB61619', 'CD86811', 'RU83859', 'FG63582', 'NN71951', 'WB37082', 'SM52139', 'FL82372', 'DP45816', 'GW33762', 'RZ33670', 'PY70169', 'MO91628', 'HW87852', 'HB20453', 'BN87372', 'YX23800', 'DZ87709', 'XW13033', 'SP81997', 'OM82309', 'ZU35962', 'VH85817', 'DT85712', 'YJ88573', 'SQ1

1. Dataset has 4008 rows and 11 columns.
2. Data types mostly appropriate; two columns were fixed.
3. Several categorical fields need normalization:
   - GENDER: unify casing and spelling.
   - ST: standardize abbreviations.
   - Education: merge similar names.
4. Some numeric fields (Monthly Premium Auto, Lifetime Value) show potential outliers.
5. Data is ready for deeper analysis after cleanup.

## Challenge 2: analyzing the data

### Exercise 1

The marketing team wants to know the top 5 less common customer locations. Create a pandas Series object that contains the customer locations and their frequencies, and then retrieve the top 5 less common locations in ascending order.

In [23]:
# Your code here

import pandas as pd

# Create a pandas Series of customer locations and their frequencies
location_freq = data['ST'].value_counts()

# Retrieve the 5 least common locations in ascending order
least_common_5 = location_freq.sort_values(ascending=True).head(5)

# Display results
print("Top 5 Least Common Customer Locations:\n")
print(least_common_5)


Top 5 Least Common Customer Locations:

ST
AZ             25
WA             30
Washington     81
Nevada         98
Cali          120
Name: count, dtype: int64


### Exercise 2

The sales team wants to know the total number of policies sold for each type of policy. Create a pandas Series object that contains the policy types and their total number of policies sold, and then retrieve the policy type with the highest number of policies sold.

*Hint:*
- *Using value_counts() method simplifies this analysis.*
- *Futhermore, there is a method that returns the index of the maximum value in a column or row.*


In [25]:
# Your code here

# Create a Series showing total number of policies sold per type
policy_counts = data['Policy Type'].value_counts()

# Retrieve the policy type with the highest number of policies sold
top_policy_type = policy_counts.idxmax()
top_policy_count = policy_counts.max()

# Display the results
print("Total Policies Sold by Type:\n")
print(policy_counts)
print("\nPolicy Type with the Highest Sales:")
print(f"{top_policy_type} → {top_policy_count} policies sold")


Total Policies Sold by Type:

Policy Type
Personal Auto     780
Corporate Auto    234
Special Auto       57
Name: count, dtype: int64

Policy Type with the Highest Sales:
Personal Auto → 780 policies sold


### Exercise 3

The sales team wants to know if customers with Personal Auto have a lower income than those with Corporate Auto. How does the average income compare between the two policy types?

- Use *loc* to create two dataframes: one containing only Personal Auto policies and one containing only Corporate Auto policies.
- Calculate the average income for each policy.
- Print the results.

In [26]:
# Your code here

import pandas as pd

# Create separate DataFrames for each policy type using .loc

personal_auto = data.loc[data['Policy Type'] == 'Personal Auto']
corporate_auto = data.loc[data['Policy Type'] == 'Corporate Auto']

# Calculate average income for each policy type

avg_income_personal = personal_auto['Income'].mean()
avg_income_corporate = corporate_auto['Income'].mean()

# Display the results

print("Average Income by Policy Type:\n")
print(f"Personal Auto:  ${avg_income_personal:,.2f}")
print(f"Corporate Auto: ${avg_income_corporate:,.2f}")

# Optional: Simple conclusion
if avg_income_personal < avg_income_corporate:
    print("\nCustomers with Personal Auto policies have a LOWER average income.")
else:
    print("\nCustomers with Personal Auto policies have a HIGHER average income.")


Average Income by Policy Type:

Personal Auto:  $38,180.70
Corporate Auto: $41,390.31

Customers with Personal Auto policies have a LOWER average income.


### Bonus: Exercise 4


Your goal is to identify customers with a high policy claim amount.

Instructions:

- Review again the statistics for total claim amount to gain an understanding of the data.
- To identify potential areas for improving customer retention and profitability, we want to focus on customers with a high policy claim amount. Consider customers with a high policy claim amount to be those in the top 25% of the total claim amount. Create a pandas DataFrame object that contains information about customers with a policy claim amount greater than the 75th percentile.
- Use DataFrame methods to calculate summary statistics about the high policy claim amount data. 

*Note: When analyzing data, we often want to focus on certain groups of values to gain insights. Percentiles are a useful tool to help us define these groups. A percentile is a measure that tells us what percentage of values in a dataset are below a certain value. For example, the 75th percentile represents the value below which 75% of the data falls. Similarly, the 25th percentile represents the value below which 25% of the data falls. When we talk about the top 25%, we are referring to the values that fall above the 75th percentile, which represent the top quarter of the data. On the other hand, when we talk about the bottom 25%, we are referring to the values that fall below the 25th percentile, which represent the bottom quarter of the data. By focusing on these groups, we can identify patterns and trends that may be useful for making decisions and taking action.*

*Hint: look for a method that gives you the percentile or quantile 0.75 and 0.25 for a Pandas Series.*

*Hint 2: check `Boolean selection according to the values of a single column` in https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9*

In [17]:
# Your code here